### PR TITLE
Implement precompiled queries

### LINKED ISSUE #31331: Use interceptors to redirect the query to precompiled code
See https://github.com/dotnet/roslyn/blob/main/docs/features/interceptors.md

* [ ] Identify and properly intercept all terminating operators
    * [ ] ExecuteUpdate/Delete
* [ ] Use ISyntaxGenerator in LinqToCSharpSyntaxGenerator, to avoid e.g. parentheses/operator precedence issues. 

Specific query types to keep in mind and test:

* [ ] SQL queries
* [ ] Split query
* [ ] GroupBy final operator (materializer work)
* [ ] Shared-type entity types (in various places where we do entity lookups)
* [ ] Improve tests to assert on results, possibly on SQL.

### PR DESCRIPTION
**Pull request by roji**
The precompiled query work is in good enough shape that I think it's the right moment to submit as a draft PR. There are still various test failures, checks and cleanups which I plan to implement, but it's more about edge cases at this point - the overall architecture should be sound. I'm happy to do a session with whoever wants to dive in and at least explain the general approach (we can also do a brown bag at some point if people want to).

Notes:

* This PR is based on top of #32721 (liftable constants in shaper), which @maumar is working on (**take care to filter out the 1st commit when reviewing**). The shaper work must be complete for this PR to be completed.
* One main reason to submit this now, is to allow @AndriySvyryd to react:
    * LinqToCSharpSyntaxTranslator now generates unsafe accessors when it encounters private field/property accesses; @AndriySvyryd I think this can be used instead of the previous "method access replacements" approach for compiled model generation, where the unsafe accessors are pre-generated externally and then substituted in the overriding implementation.
    * Work can begin on integrating this into MSBuild.
* This is a pretty massive PR, implementing the bulk (and probably most complex parts) of precompiled queries; the more eyes on this, the better.

Closes #31331

### CODE REVIEW

File: src/EFCore.Design/Query/Internal/PrecompiledQueryCodeGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,1362 @@
...
+        _csharpToLinqTranslator = new CSharpToLinqTranslator();
+    }
+
+    /// <summary>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    public async Task GeneratePrecompiledQueries(
```
Comment (roji) [inline src/EFCore.Design/Query/Internal/PrecompiledQueryCodeGenerator.cs:65]: Mainly for @AndriySvyryd: this method is currently just hacked up to be invoked from e.g. an external console program, and needs to be adapted (in case it's not obvious). The precompiled query tests I've been working with call the method below, which just accepts a Compilation directly.

File: src/EFCore.Design/Scaffolding/Internal/CSharpRuntimeModelCodeGenerator.cs [hunk 2]
```diff
@@ -1824,30 +1827,33 @@ private void
             out var createAndSetCollection,
             out var createCollection);
 
+        // TODO
+        var unsafeAccessors = new HashSet<string>();
```
Comment (roji) [inline src/EFCore.Design/Scaffolding/Internal/CSharpRuntimeModelCodeGenerator.cs:1831]: @AndriySvyryd note the unsafe accessors now getting generated inside LinqToCSharpSyntaxTranslators - ideally these would be used here instead of the current method replacements mechanism?

Comment (AndriySvyryd) [reply src/EFCore.Design/Scaffolding/Internal/CSharpRuntimeModelCodeGenerator.cs:1831]: There are two reasons why I implemented them differently:

1. Avoid generating duplicate accessors for the same members
2. Support private property access

Comment (roji) [reply src/EFCore.Design/Scaffolding/Internal/CSharpRuntimeModelCodeGenerator.cs:1831]: For the deduplication, this already occurs within the translator itself, so we could expose what's needed to allow deduplication across multiple invocations of the translator. For example, we could have the same instance of the translator accumulate unsafe accessor declarations internally as its reused, and at the end you'd all of them via a special API. Or we could externalize that state so you'd pass it back in, etc.

(For that state, originally I had a nice and clean dictionary from MemberInfo to the MethodDeclarationSyntax representing its interceptor - but needing different interceptors for field read/write makes that a tiny bit more complex.)

Private property access should also already be implemented in the translator - let me know if you see an issue.

Comment (AndriySvyryd) [reply src/EFCore.Design/Scaffolding/Internal/CSharpRuntimeModelCodeGenerator.cs:1831]: Using the same instance won't help when processing an assembly that doesn't contain the compiled model.

I was thinking that we can store `memberAccessReplacements` as an annotation in the compiled model and precompiled queries would use it. If you think that it's possible to have queries that access a private member that's not in the model then we can also keep `unsafeAccessors`

Comment (roji) [reply src/EFCore.Design/Scaffolding/Internal/CSharpRuntimeModelCodeGenerator.cs:1831]: So FWIW, I looked again and the current implementation already deduplicates unsafe accessors across invocations (i.e. across shapers of different queries) - the state passed around is simply `ISet<MethodDeclarationSyntax>`, and at the end we dump that out into the generated file.

> Using the same instance won't help when processing an assembly that doesn't contain the compiled model.

Let's do a call so I can understand this better (possibly tomorrow before/after the design meeting?)

File: src/EFCore.Design/Query/Internal/PrecompiledQueryCodeGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,1362 @@
...
+                .Append("System.Reflection")
+                .Append("System.Collections.Generic")
+                .Append("Microsoft.EntityFrameworkCore")
+                .Append("Microsoft.EntityFrameworkCore.Query")
+                .Append("Microsoft.EntityFrameworkCore.ChangeTracking.Internal")
+                .Append("Microsoft.EntityFrameworkCore.Query.Internal")
+                .Append("Microsoft.EntityFrameworkCore.Diagnostics")
+                .Append("Microsoft.EntityFrameworkCore.Infrastructure")
+                .Append("Microsoft.EntityFrameworkCore.Infrastructure.Internal")
+                .Append("Microsoft.EntityFrameworkCore.Metadata")
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Query/Internal/PrecompiledQueryCodeGenerator.cs:305]: Add these to _namespaces

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.Helper.cs [hunk 4]
```diff
@@ -595,14 +595,14 @@ protected override Expression VisitExtension(Expression expression)
...
-                    selectExpression.Projection,
                     visitedTables ?? selectExpression.Tables,
                     selectExpression.Predicate,
                     selectExpression.GroupBy,
                     selectExpression.Having,
+                    selectExpression.Projection,
                     selectExpression.Orderings,
-                    selectExpression.Limit,
-                    selectExpression.Offset));
+                    selectExpression.Offset,
```
Comment (roji) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.Helper.cs:604]: Note that here - and below - I'm generally moving arround SelectExpression components to be in the order in which they're evaluated in SQL (i.e. offset before limit). We weren't always consistent, and the confusion can be hard to track down.

File: src/EFCore.Design/Scaffolding/Internal/CSharpRuntimeModelCodeGenerator.cs (hunk 2 above)
Comment (roji) [reply src/EFCore.Design/Scaffolding/Internal/CSharpRuntimeModelCodeGenerator.cs:1860]: @AndriySvyryd for now I've left the these changes as they are in the PR, let's revisit and clean up after merge OK?

**Review by roji — COMMENTED** [review]
I've more of less finished what I consider to be the most of the work here. There are some edge cases I know which aren't handled (and most probably some others I don't know), but I think it's in good enough shape to merge and fix local problems later.

So moving this out of draft state - it should be ready for a good review. Note that this is still based on top of #33351 which isn't yet done; be sure to exclude that when reviewing.

Note that

**Comment by cincuranet** [comment]
> be sure to exclude that when reviewing.

For your convenience: https://github.com/dotnet/efcore/pull/33297/files/b66d7b9c4fd5846ff6e6afa8e71dfe2f97983aa5..53025c3bfb3503257a49f2d96327e4c9ee1485db

File: src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs [hunk 5]
```diff
@@ -0,0 +1,1309 @@
...
+                originalDefinition = originalDefinition.ReducedFrom;
+            }
+
+            // To accurately find the right open generic method definition based on the Roslyn symbol, we need to create a mapping between
+            // generic type parameter names (based on the Roslyn side) and .NET reflection Types representing those type parameters.
+            // This includes both type parameters immediately on the generic method, as well as type parameters from the method's
+            // containing type (and recursively, its containing types)
+            var typeTypeParameterMap = new Dictionary<string, Type>(Foo(methodSymbol.ContainingType));
+
+            IEnumerable<KeyValuePair<string, Type>> Foo(INamedTypeSymbol typeSymbol)
```
Comment (cincuranet) [inline src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs:545]: I know it is a "private" helper, but maybe better name than `Foo`?

File: src/EFCore.Design/Query/Internal/QueryLocator.cs [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+        CancellationToken cancellationToken = default)
+    {
+        if (_compilation is null)
+        {
+            throw new InvalidOperationException("A compilation must be loaded.");
+        }
+
+        if (!_compilation.SyntaxTrees.Contains(syntaxTree))
+        {
+            throw new ArgumentException("");
```
Comment (cincuranet) [inline src/EFCore.Design/Query/Internal/QueryLocator.cs:80]: Missing message/argument.

File: test/EFCore.Design.Tests/EFCore.Design.Tests.csproj [hunk 7]
```diff
@@ -56,7 +56,7 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="$(MicrosoftCodeAnalysisVersion)" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.10.0-2.final" />
```
Comment (cincuranet) [inline test/EFCore.Design.Tests/EFCore.Design.Tests.csproj:59]: This should be probably fixed back before merging.

File: test/EFCore.Relational.Specification.Tests/EFCore.Relational.Specification.Tests.csproj [hunk 8]
```diff
@@ -48,7 +48,11 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="$(MicrosoftCodeAnalysisVersion)" />
+    <!-- Newer version of Roslyn used in tests for testing interceptors -->
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.10.0-2.final" />
```
Comment (cincuranet) [inline test/EFCore.Relational.Specification.Tests/EFCore.Relational.Specification.Tests.csproj:52]: Probably fix before merging.

File: test/EFCore.SqlServer.FunctionalTests/Scaffolding/Baselines/Custom_function_parameter_type_mapping/FunctionParameterTypeMappingContextModelBuilder.cs [hunk 9]
```diff
@@ -50,15 +50,15 @@ private FunctionParameterTypeMappingContextModel()
             param.TypeMapping = StringTypeMapping.Default.Clone(
                 comparer: new ValueComparer<string>(
                     (string v1, string v2) => v1 == v2,
-                    (string v) => v.GetHashCode(),
+                    (string v) => ((object)v).GetHashCode(),
```
Comment (cincuranet) [inline test/EFCore.SqlServer.FunctionalTests/Scaffolding/Baselines/Custom_function_parameter_type_mapping/FunctionParameterTypeMappingContextModelBuilder.cs:53]: For my own education, why this cast?

File: test/EFCore.Design.Tests/EFCore.Design.Tests.csproj (hunk 7 above)
Comment (roji) [reply test/EFCore.Design.Tests/EFCore.Design.Tests.csproj:59]: This version contains an interceptor-related compiler fix which is required for the tests to run successfully - so I don't think we'll be able to change this back. However, it's a test-only dependency on a preview package, and 4.10.0 should release for 9.0, at which point hopefully we can align everything.

File: test/EFCore.Relational.Specification.Tests/EFCore.Relational.Specification.Tests.csproj (hunk 8 above)
Comment (roji) [reply test/EFCore.Relational.Specification.Tests/EFCore.Relational.Specification.Tests.csproj:52]: See https://github.com/dotnet/efcore/pull/33297#discussion_r1558066893

File: test/EFCore.SqlServer.FunctionalTests/Scaffolding/Baselines/Custom_function_parameter_type_mapping/FunctionParameterTypeMappingContextModelBuilder.cs (hunk 9 above)
Comment (roji) [reply test/EFCore.SqlServer.FunctionalTests/Scaffolding/Baselines/Custom_function_parameter_type_mapping/FunctionParameterTypeMappingContextModelBuilder.cs:53]: This code is a result of LINQ->C# translation, which is complicated business 😅 ([here's the translator](https://github.com/dotnet/efcore/blob/main/src/EFCore.Design/Query/Internal/RuntimeModelLinqToCSharpSyntaxTranslator.cs)).

Basically there was a case where materializer LINQ code was calling the [`ICollection<T>.IsReadOnly`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.icollection-1.isreadonly?view=net-7.0#system-collections-generic-icollection-1-isreadonly) property on a `List<int>`. Unfortunately, doing that in simple C# code (`new List<int>().IsReadOnly`) is illegal since IsReadOnly is implemented as an explicit interface implementation; so I modified the LINQ->C# translator to emit an explicit up-cast whenever the property/method being called is declared on another type as the expression itself. That has the unfortunate consequence of adding the cast above, since GetHashCode() is implemented on `object`, not on `string`.

It should definitely be possible to make a more accurate/narrow check to specifically detect explicit interface implementations, and only emit the cast then - but I didn't go that far for now.

/cc @maumar for the fun aspects of LINQ->C#...

File: src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs (hunk 5 above)
Comment (roji) [reply src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs:545]: Yeah, some leftovers from the early prototype hacking days ;)

File: test/EFCore.SqlServer.FunctionalTests/Scaffolding/Baselines/Custom_function_parameter_type_mapping/FunctionParameterTypeMappingContextModelBuilder.cs (hunk 9 above)
Comment (cincuranet) [reply test/EFCore.SqlServer.FunctionalTests/Scaffolding/Baselines/Custom_function_parameter_type_mapping/FunctionParameterTypeMappingContextModelBuilder.cs:53]: Thanks.

File: test/EFCore.Design.Tests/EFCore.Design.Tests.csproj (hunk 7 above)
Comment (cincuranet) [reply test/EFCore.Design.Tests/EFCore.Design.Tests.csproj:59]: I guess, we can change the `MicrosoftCodeAnalysisVersion` variable. For 9.0 RTM everything will align fine.

File: src/EFCore/Query/ExpressionPrinter.cs [hunk 10]
```diff
@@ -461,13 +461,19 @@ protected override Expression VisitConditional(ConditionalExpression conditional
...
-        }
-        else
-        {
-            PrintValue(constantExpression.Value);
+            case IPrintableExpression printable:
+                printable.Print(this);
+                break;
+
+            case IQueryable queryable:
+                _stringBuilder.Append(Print(queryable.Expression));
```
Comment (maumar) [inline src/EFCore/Query/ExpressionPrinter.cs:471]: Visit(queryable.Expression)

File: src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs [hunk 11]
```diff
@@ -0,0 +1,1309 @@
...
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </remarks>
+    public virtual Expression Translate(SyntaxNode node, SemanticModel semanticModel)
+    {
+        if (_compilation is null)
+        {
+            throw new InvalidOperationException("A compilation must be loaded.");
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs:94]: Move to resources

File: src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs [hunk 12]
```diff
@@ -0,0 +1,1309 @@
...
+        _capturedVariables.Clear();
+        foreach (var captured in _semanticModel.AnalyzeDataFlow(node).Captured)
+        {
+            _capturedVariables[captured] = null;
+        }
+
+        var result = Visit(node);
+
+        // TODO: Sanity check: make sure all captured variables in _capturedVariables have non-null values
+        // (i.e. have been encountered and referenced)
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs:113]: Add an issue # to all `TODO`. Could be just a checklist of everything left to do.

File: src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs [hunk 13]
```diff
@@ -0,0 +1,1309 @@
...
+            SyntaxKind.MultiplyExpression => Multiply(left, right),
+            SyntaxKind.DivideExpression => Divide(left, right),
+            SyntaxKind.ModuloExpression => Modulo(left, right),
+            SyntaxKind.LeftShiftExpression => LeftShift(left, right),
+            SyntaxKind.RightShiftExpression => RightShift(left, right),
+            // TODO UnsignedRightShiftExpression
+            SyntaxKind.LogicalOrExpression => OrElse(left, right),
+            SyntaxKind.LogicalAndExpression => AndAlso(left, right),
+
+            // For bitwise operations over enums, we the enum to its underlying type before the bitwise operation, and then back to the
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs:253]:
```suggestion
            // For bitwise operations over enums, we cast the enum to its underlying type before the bitwise operation, and then back to the
```

File: src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs [hunk 14]
```diff
@@ -0,0 +1,1309 @@
...
+                        t => t.IndexParameters.Length == arguments.Count
+                            && t.IndexParameters
+                                .Select(p => p.ParameterType)
+                                .SequenceEqual(arguments.Select(a => ResolveType(a.Expression))))
+                    .Select(t => t.Property)
+                    .FirstOrDefault();
+
+                if (property?.GetMethod is null)
+                {
+                    throw new UnreachableException("No matching property found for ElementAccessExpressionSyntax");
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs:339]: Use `Check.DebugAssert` instead

File: src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs [hunk 15]
```diff
@@ -0,0 +1,1309 @@
...
+                originalDefinition = originalDefinition.ReducedFrom;
+            }
+
+            // To accurately find the right open generic method definition based on the Roslyn symbol, we need to create a mapping between
+            // generic type parameter names (based on the Roslyn side) and .NET reflection Types representing those type parameters.
+            // This includes both type parameters immediately on the generic method, as well as type parameters from the method's
+            // containing type (and recursively, its containing types)
+            var typeTypeParameterMap = new Dictionary<string, Type>(GetTypeTypeParameters(methodSymbol.ContainingType));
+
+            IEnumerable<KeyValuePair<string, Type>> GetTypeTypeParameters(INamedTypeSymbol typeSymbol)
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs:545]: Move to the end and make static

File: src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs [hunk 16]
```diff
@@ -0,0 +1,1309 @@
...
+    /// </remarks>
+    public virtual Type TranslateType(ITypeSymbol typeSymbol)
+        => ResolveType(typeSymbol);
+
+    private Type ResolveType(SyntaxNode node)
+        => _semanticModel.GetTypeInfo(node).Type is { } typeSymbol
+            ? ResolveType(typeSymbol)
+            : throw new InvalidOperationException("Could not find type symbol for: " + node);
+
+    private Type ResolveType(ITypeSymbol typeSymbol, Dictionary<string, Type>? genericParameterMap = null)
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs:1069]: Review whether more calls of this should be done while supplying a `genericParameterMap`. For example, the expression could be referencing a generic type parameter from the containing type.

File: src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs [hunk 17]
```diff
@@ -1565,18 +1568,27 @@ protected override Expression VisitMember(MemberExpression member)
...
-                        : Translate<ExpressionSyntax>(member.Expression),
-                    IdentifierName(member.Member.Name));
+                var expression = member switch
+                {
+                    // Static member
+                    { Expression: null } => Generate(member.Member.DeclaringType!),
+
+                    // If the member isn't declared on the same type as the expression, (e.g. explicit interface implementation), add
+                    // a cast up to the declaring type.
+                    _ when member.Member.DeclaringType is Type declaringType && declaringType != member.Expression.Type
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs:1584]: Add a check whether the declaring type is an interface or the member is hidden

File: src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs [hunk 18]
```diff
@@ -1868,9 +2055,9 @@ protected override Expression VisitNew(NewExpression node)
                 initializer: null);
         }
 
-        if (node.Constructor?.DeclaringType is not null)
+        if (node.Type.Namespace is not null)
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs:2058]: This check is redundant

File: src/EFCore.Design/Query/Internal/PrecompiledQueryCodeGenerator.cs [hunk 19]
```diff
@@ -0,0 +1,1132 @@
...
+        }
+
+        _code.AppendLine(mainCode);
+
+        _code.AppendLine(
+            """
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+    sealed class InterceptsLocationAttribute : Attribute
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Query/Internal/PrecompiledQueryCodeGenerator.cs:285]:
```suggestion
    internal sealed class InterceptsLocationAttribute : Attribute
```

File: test/EFCore.Design.Tests/Query/CSharpToLinqTranslatorTest.cs [hunk 20]
```diff
@@ -0,0 +1,537 @@
...
+using static System.Linq.Expressions.Expression;
+
+namespace Microsoft.EntityFrameworkCore.Query;
+
+// ReSharper disable InconsistentNaming
+// ReSharper disable RedundantCast
+
+#nullable enable
+
+public class CSharpToLinqTranslatorTest
```
Comment (AndriySvyryd) [inline test/EFCore.Design.Tests/Query/CSharpToLinqTranslatorTest.cs:18]: Consider merging with `LinqToCSharpSyntaxTranslatorTest` as roundtrip tests

File: test/EFCore.Design.Tests/Query/CSharpToLinqTranslatorTest.cs [hunk 21]
```diff
@@ -0,0 +1,537 @@
...
+            "DateTime.Now");
+
+    // TODO: MemberAccess on fields
+
+    [Fact]
+    public void Nested_type()
+        => AssertExpression(
+            () => (object)new Blog(),
+            "(object)new CSharpToLinqTranslatorTest.Blog()");
+
```
Comment (AndriySvyryd) [inline test/EFCore.Design.Tests/Query/CSharpToLinqTranslatorTest.cs:331]: Add tests for anonymous type, different closures, nested generics, indexers, records `with`

File: src/EFCore/Properties/CoreStrings.resx [hunk 22]
```diff
@@ -1333,6 +1339,9 @@
   <data name="PoolingOptionsModified" xml:space="preserve">
     <value>'OnConfiguring' cannot be used to modify DbContextOptions when DbContext pooling is enabled.</value>
   </data>
+  <data name="PrecompiledQueryNotSupported" xml:space="preserve">
+    <value>Precompiled queries aren't supported by your EF provider.</value>
```
Comment (AndriySvyryd) [inline src/EFCore/Properties/CoreStrings.resx:1343]:
```suggestion
    <value>Precompiled queries aren't supported by the current EF provider.</value>
```

**Review by AndriySvyryd — APPROVED** [review]

File: test/EFCore.Design.Tests/EFCore.Design.Tests.csproj (hunk 7 above)
Comment (roji) [reply test/EFCore.Design.Tests/EFCore.Design.Tests.csproj:59]: We can, but this would limit the SDK versions that can be used by ordinary users with EF... We've had issues with changes like this in the past - given that we only need this in testing, I'd rather we left things as-is for now and update later as needed...

File: src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs (hunk 12 above)
Comment (roji) [reply src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs:113]: #33639

File: src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs (hunk 15 above)
Comment (roji) [reply src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs:545]: I don't think this can be static, since it invokes ResolveType... But moving to the end.

File: src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs (hunk 16 above)
Comment (roji) [reply src/EFCore.Design/Query/Internal/CSharpToLinqTranslator.cs:1069]: Yeah, I suspect there are a few bugs lurking around this... With everything going on, I'll defer this to now, and have added a note in #33639. We can fix later based on feedback or do a more thorough review.

File: src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs (hunk 17 above)
Comment (roji) [reply src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs:1584]: Changed to only add the parentheses for interface-declared members - but what do you have in mind for hidden?

File: src/EFCore.Design/Query/Internal/PrecompiledQueryCodeGenerator.cs (hunk 19 above)
Comment (roji) [reply src/EFCore.Design/Query/Internal/PrecompiledQueryCodeGenerator.cs:285]: I think it should actually be `file`-scoped...

File: test/EFCore.Design.Tests/Query/CSharpToLinqTranslatorTest.cs (hunk 20 above)
Comment (roji) [reply test/EFCore.Design.Tests/Query/CSharpToLinqTranslatorTest.cs:18]: IMHO it leaves things simpler for each test/test class to focus on one component here, given the complexity of the components...

File: test/EFCore.Design.Tests/Query/CSharpToLinqTranslatorTest.cs (hunk 21 above)
Comment (roji) [reply test/EFCore.Design.Tests/Query/CSharpToLinqTranslatorTest.cs:331]: Added note in #33639. I hope we get to do a good bash on this before the release, and flush out bugs.

File: src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs (hunk 17 above)
Comment (AndriySvyryd) [reply src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs]: It's a corner case, but if the current type hides the member being used then we'd need to add a cast to the declaring type to be sure that the same member is used

Comment (roji) [reply src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs]: Makes sense - added note in 33639.
